enhancements/network/on-prem-mutable-vips.md

```diff
@@ -0,0 +1,199 @@
+
+## Design Details
+
+This feature will require changes in a few different components:
+
+* Installer - New fields will need to be populated during the installation. This is a safe part as installer is already populating `status` and now will populate `status` and `spec`.
+
+* API - New fields will need to be added to the platform spec section of the infrastructure object.
+
+* Cluster Network Operator - New object will need to be watched by one of its controllers. The core logic of validating the requested change will be implemented there.
+
+* Machine Config Operator - Definition of the runtimecfg Pod will need to be rendered again when the VIP configuration changes.
+
+* baremetal-runtimecfg - The code rendering the keepalived template belongs to this repo. As the enhancement does not touch the rendering part no big changes are expected in the runtimecfg. This component is used to re-render configuration in case something changes in the cluster (e.g. new nodes are added) so the fact we modify configuration is not a new feature from its perspective. Small changes may be needed to acommodate to the new user scenario to provide good user experience.
+
+### openshift/api
+
+The new structure of platformSpec would look like this:
+
+```yaml
+platformSpec:
+  properties:
+    baremetal:
+      description: BareMetal contains settings specific to the BareMetal platform.
+      type: object
+      properties:
+        apiServerInternalIPs:
+          description: apiServerInternalIPs are the IP addresses to...
+          type: array
+        ingressIPs:
+          description: ingressIPs are the external IPs which...
+          type: array
+      ...
+```
+
+This format (i.e. usage of arrays) is already used by platformStatus so this does not introduce a new type nor schema.
+
+### Cluster Network Operator
+
+There already exists `infrastructureconfig_controller` inside CNO that watches for changes in `configv1.Infrastructure` for on-prem platforms. Because of this we do not need to create a new one nor change any fundaments, we only need to extend its logic.
+
+Currently the controller already reconciles on the `CreateEvent` and `UpdateEvent` so we will implement a set of functions that compare and validate `platformSpec` with `platformStatus` and do what's needed.
+
+### Machine Config Operator
+
+[Keepalived templates managed by the Machine Config Operator](https://github.com/openshift/machine-config-operator/blob/2b31b5b58f0d7d9fe6c3e331e4b3f01c9a1bd00c/templates/common/on-prem/files/keepalived.yaml#L46-L49) currently use the `PlatformStatus` fields. We will need to change the code inside the controller which [pulls this value](https://github.com/openshift/machine-config-operator/blob/2b31b5b58f0d7d9fe6c3e331e4b3f01c9a1bd00c/pkg/controller/template/render.go#L551-L575) so that new `PlatformSpec` path is used.
```

Comment:
So we'll be ignoring platformStatus for these fields after this change? I suspect we're going to be asked to sync the spec and status fields. Maybe we just want CNO to synchronize the changes in spec to status?

We had discussed having $SOMETHING update the status only after the changes are complete, but I have to admit I'm not sure how that will work when CNO is responsible for updating Infrastructure but MCO is responsible for actually rolling out the changes.

Reply:
My idea was that CNO becomes responsible and CNO syncs `status` with `spec` after user modifies the Infrastructure CR. If the change passed validations and is rolled-out, CNO would set the `status` to what you want. If the change is rejected, it would not.

Reply:
I wrote it explicitly that today PlatformStatus is not synced from any real-time runtime monitoring and that I wish to keep it that way. It's not ideal but the elegant way to do it would be to have a controller that monitors keepalived at runtime, but this would be a lot of work that is probably not something we want to do now

Reply:
Fair enough. It's a good point that platformStatus isn't necessarily indicative that a given configuration has been applied today either.